include/eve/detail/function/simd/x86/swizzle.hpp

```diff
@@ -0,0 +1,258 @@
+        auto const vv     = bit_cast(v,as_<f_t>{});
+        constexpr auto m  = _MM_SHUFFLE2(q(1,2)&1, q(0,2)&1);
+        return bit_cast(process_zeros(f_t{_mm_shuffle_pd(vv,vv,m)},q),as_<that_t>{});
+      }
+      else if constexpr(sizeof(T) == 4)
+      {
+        constexpr auto m  = _MM_SHUFFLE(q(3,4)&3, q(2,4)&3, q(1,4)&3, q(0,4)&3);
+        if constexpr( std::same_as<T,float> ) return process_zeros(that_t{_mm_shuffle_ps(v,v,m)} ,q);
+        else                                  return process_zeros(that_t{_mm_shuffle_epi32(v,m)},q);
+      }
+      else if constexpr(sizeof(T) == 2)
+      {
+        if constexpr( (sz < 8) && (q < 4) )
+        {
+          constexpr auto m  = _MM_SHUFFLE(q(3,4)&3, q(2,4)&3, q(1,4)&3, q(0,4)&3);
+          return process_zeros(that_t{_mm_shufflelo_epi16(v,m)},q);
+        }
+        else if constexpr( (sz < 8) && (q >= 4) )
+        {
+          constexpr auto m  = _MM_SHUFFLE(q(3,4)&3, q(2,4)&3, q(1,4)&3, q(0,4)&3);
+          return process_zeros(that_t{_mm_bsrli_si128(_mm_shufflehi_epi16(v,m),8)},q);
+        }
+        else
+        {
+          constexpr auto lp = pattern_view<0,4,8>(q);
+          constexpr auto hp = pattern_view<4,8,8>(q);
+
+                if constexpr( lp < 4 && hp >= 4) return process_zeros(that_t{v[lp],v[hp]},q);
+          else  if constexpr( hp < 4 && lp >= 4) return process_zeros(that_t{v[lp],v[hp]},q);
+          else  return basic_swizzle_(EVE_RETARGET(cpu_),v,q);
+        }
+      }
+      else
+      {
+        return basic_swizzle_(EVE_RETARGET(cpu_),v,q);
+      }
+    }
+  }
+
+  //================================================================================================
+  // AVX+ variant
+  //================================================================================================
+  template<typename T, typename N, x86_abi ABI, shuffle_pattern Pattern>
+  EVE_FORCEINLINE auto basic_swizzle_( EVE_SUPPORTS(avx_), wide<T,N,ABI> const& v, Pattern const& p)
+  {
+    constexpr auto cd = N::value;
+    constexpr auto sz = Pattern::size(cd);
+    using that_t      = as_wide_t<wide<T,N,ABI>,fixed<sz>>;
+
+    constexpr auto width_in   = cd*sizeof(T);
+    constexpr auto width_out  = sz*sizeof(T);
+
+    constexpr Pattern q = {};
+
+    // We're swizzling so much we aggregate the output
+    if constexpr( has_aggregated_abi_v<that_t> )
+    {
+      return aggregate_swizzle(v,q);
+    }
+    else if constexpr( width_in == 64 )
+    {
+      /// TODO: AVX512VBMI supports 128-bits permutexvar
+      if constexpr(width_out <= 16)       return basic_swizzle_(EVE_RETARGET(cpu_),v,q);
+      else if constexpr(width_out == 32)
+      {
+              if constexpr( q <  cd/2 )  return v.slice(lower_)[ q ];
+        else  if constexpr( q >= cd/2 )  return v.slice(upper_)[ slide_pattern<cd/2,sz>(q) ];
+        /// TODO: optimize using SSSE3 + binary shuffle
+        else                             return basic_swizzle_(EVE_RETARGET(cpu_),v,q);
```

Comment:
Why retarget cpu then?

Reply:
becasue right now we don't have the shuffle for optimization